File: sdk/android/servify-sdk/src/main/kotlin/servify/sdk/android/ui/ChatPanel.kt
package servify.sdk.android.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.imePadding
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch
import servify.sdk.android.ServifyChat
import servify.sdk.android.connect.ConnectionState
import servify.sdk.android.model.SenderType

/**
 * 会话面板（D8 三形态的抽屉/全屏共体：形态差异只在宿主容器高度，主体组件同一棵）。
 * 消息流收敛经 [ChatPanelState]（纯 reducer，脱离 Compose 单测）；连接/未读等流直接接门面。
 * 视觉全部自绘（ChatText/Box/Canvas，D9 白名单，不引 material）。
 */
@Composable
fun ChatPanel(
    chat: ServifyChat,
    title: String,
    welcomeText: String?,
    onDismiss: () -> Unit,
    modifier: Modifier = Modifier,
) {
    // 面板打开即会话页可见：清未读（§4.3）。
    LaunchedEffect(chat) { chat.onSessionVisible() }

    val panelState = remember(chat) { ChatPanelState(welcomeText) }
    var version by remember(chat) { mutableIntStateOf(0) }
    val scope = rememberCoroutineScope()
    val connectionState by chat.events.connectionState.collectAsState()
    var statusLine by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(chat) {
        chat.events.messages.collect {
            panelState.onMessage(it)
            version++
        }
    }
    // 转人工状态机渲染（D8：排队提示 / 坐席接入提示，显示最近一条）。
    LaunchedEffect(chat) {
        chat.events.waitingInQueue.collect { statusLine = it }
    }
    LaunchedEffect(chat) {
        chat.events.agentAssigned.collect { statusLine = it.message }
    }

    Column(
        modifier = modifier
            .fillMaxSize()
            .background(ChatThemeDefaults.PageBackground),
    ) {
        TitleBar(title, connectionLabel(connectionState), onDismiss)
        statusLine?.let { StatusLine(it) }
        MessageList(
            items = panelState.items(),
            version = version,
            onHandoff = {
                // V1：协议无客户端转人工帧——点击即发送固定文本消息（服务端兜底链路可达）。
                scope.launch {
                    runCatching { chat.sendMessage("请转人工客服") }
                }
            },
            modifier = Modifier.weight(1f),
        )
        InputBar(
            enabled = connectionState == ConnectionState.Connected,
            onSend = { text ->
                // 发送失败（echo 超时）经 error 流暴露给宿主；面板内静默，可重发。
                scope.launch {
                    runCatching { chat.sendMessage(text) }
                }
            },
        )
    }
}

private fun connectionLabel(state: ConnectionState): String = when (state) {
    ConnectionState.Idle -> ""
    ConnectionState.Connecting -> "连接中…"
    is ConnectionState.Reconnecting -> "重新连接中…（第 ${state.attempt} 次）"
    ConnectionState.Connected -> ""
    ConnectionState.Disconnected -> "连接已断开"
}

@Composable
private fun TitleBar(title: String, status: String, onDismiss: () -> Unit) {
    val primary = LocalServifyPrimary.current
    Column(modifier = Modifier.fillMaxWidth().background(primary).statusBarsPadding()) {
        Row(
            modifier = Modifier.fillMaxWidth().height(52.dp).padding(horizontal = 16.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            ChatText(
                title,
                color = Color.White,
                fontSize = 17.sp,
                fontWeight = FontWeight.SemiBold,
                modifier = Modifier.weight(1f, fill = false),
                maxLines = 1,
            )
            if (status.isNotEmpty()) {
                Spacer(Modifier.widthIn(8.dp))
                ChatText(status, color = Color.White.copy(alpha = 0.85f), fontSize = 12.sp, maxLines = 1)
            }
            Spacer(Modifier.weight(1f))
            ChatText(
                "✕",
                color = Color.White,
                fontSize = 18.sp,
                modifier = Modifier
                    .clip(RoundedCornerShape(8.dp))
                    .clickable(onClick = onDismiss)
                    .padding(6.dp),
            )
        }
    }
}

@Composable
private fun StatusLine(text: String) {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .background(ChatThemeDefaults.HandoffBadge)
            .padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        ChatText(text, color = ChatThemeDefaults.TextPrimary, fontSize = 13.sp)
    }
}

@Composable
private fun MessageList(
    items: List<ChatListItem>,
    version: Int,
    onHandoff: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val listState = rememberLazyListState()
    // 新消息（或流式增量）到达即滚到底。
    LaunchedEffect(version) {
        if (items.isNotEmpty()) listState.animateScrollToItem(items.size - 1)
    }
    LazyColumn(state = listState, modifier = modifier.fillMaxWidth()) {
        items(items, key = { item -> item.key }) { item ->
            when (item) {
                is ChatListItem.Welcome -> WelcomeItem(item)
                is ChatListItem.Bubble -> MessageBubble(item, onHandoff)
            }
        }
    }
}

private val ChatListItem.key: String
    get() = when (this) {
        is ChatListItem.Welcome -> "__welcome"
        is ChatListItem.Bubble -> id
    }

@Composable
private fun WelcomeItem(item: ChatListItem.Welcome) {
    Box(modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp)) {
        ChatText(
            item.text,
            color = ChatThemeDefaults.TextSecondary,
            fontSize = 13.sp,
            modifier = Modifier
                .clip(RoundedCornerShape(12.dp))
                .background(ChatThemeDefaults.SurfaceWhite)
                .padding(horizontal = 12.dp, vertical = 8.dp),
        )
    }
}

@Composable
private fun MessageBubble(item: ChatListItem.Bubble, onHandoff: () -> Unit) {
    val right = item.sender.isRightAligned()
    Row(
        modifier = Modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 4.dp),
        horizontalArrangement = if (right) Arrangement.End else Arrangement.Start,
    ) {
        Column(
            horizontalAlignment = if (right) Alignment.End else Alignment.Start,
            modifier = Modifier.widthIn(max = 280.dp),
        ) {
            if (item.isAiResponse && !right) {
                ChatText(
                    "AI 助手",
                    color = ChatThemeDefaults.TextSecondary,
                    fontSize = 11.sp,
                    modifier = Modifier.padding(bottom = 2.dp, start = 4.dp),
                )
            }
            Box(
                modifier = Modifier
                    .clip(RoundedCornerShape(12.dp))
                    .background(
                        if (right) ChatThemeDefaults.BubbleCustomer else ChatThemeDefaults.BubbleOther,
                    )
                    .padding(horizontal = 12.dp, vertical = 8.dp),
            ) {
                ChatText(
                    if (item.isStreaming) item.content + "…" else item.content,
                    color = if (right) Color.White else ChatThemeDefaults.TextPrimary,
                    fontSize = 15.sp,
                )
            }
            if (item.suggestsHandoff) {
                HandoffButton(onHandoff)
            }
            if (item.sources.isNotEmpty()) {
                SourcesList(item.sources)
            }
        }
    }
}

/** 置信门"转人工"强调按钮（D8：next_action=handoff 驱动）。 */
@Composable
private fun HandoffButton(onClick: () -> Unit) {
    val primary = LocalServifyPrimary.current
    Box(
        modifier = Modifier
            .padding(top = 4.dp)
            .clip(RoundedCornerShape(8.dp))
            .background(primary)
            .clickable(onClick = onClick)
            .padding(horizontal = 12.dp, vertical = 6.dp),
    ) {
        ChatText("转人工", color = Color.White, fontSize = 13.sp, fontWeight = FontWeight.Medium)
    }
}

/** 引用来源列表（D8：按得分降序的展开列表，score 已在映射层排好、null 归底）。 */
@Composable
private fun SourcesList(sources: List<SourceItem>) {
    Column(modifier = Modifier.padding(top = 4.dp)) {
        ChatText("参考来源", color = ChatThemeDefaults.TextSecondary, fontSize = 11.sp)
        sources.forEach { source ->
            ChatText(
                buildString {
                    append("· ")
                    append(source.title)
                    append("（")
                    append(source.documentId)
                    source.score?.let { append(String.format("%.2f", it)) }
                    append("）")
                },
                color = ChatThemeDefaults.TextSecondary,
                fontSize = 12.sp,
                maxLines = 1,
                modifier = Modifier.padding(top = 2.dp),
            )
        }
    }
}

@Composable
private fun InputBar(enabled: Boolean, onSend: (String) -> Unit) {
    var text by remember { mutableStateOf("") }
    val primary = LocalServifyPrimary.current
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .background(ChatThemeDefaults.SurfaceWhite)
            .navigationBarsPadding()
            .imePadding()
            .padding(horizontal = 12.dp, vertical = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Box(
            modifier = Modifier
                .weight(1f)
                .clip(RoundedCornerShape(20.dp))
                .background(ChatThemeDefaults.PageBackground)
                .padding(horizontal = 12.dp, vertical = 8.dp),
        ) {
            BasicTextField(
                value = text,
                onValueChange = { text = it },
                textStyle = TextStyle(
                    color = ChatThemeDefaults.TextPrimary,
                    fontSize = 15.sp,
                ),
                cursorBrush = SolidColor(primary),
                maxLines = 4,
                modifier = Modifier.fillMaxWidth(),
                decorationBox = { inner ->
                    if (text.isEmpty()) {
                        ChatText("输入消息…", color = ChatThemeDefaults.TextSecondary, fontSize = 15.sp)
                    }
                    inner()
                },
            )
        }
        Spacer(Modifier.widthIn(8.dp))
        val canSend = enabled && text.isNotBlank()
        Box(
            modifier = Modifier
                .clip(RoundedCornerShape(20.dp))
                .background(if (canSend) primary else ChatThemeDefaults.Divider)
                .clickable(enabled = canSend) {
                    onSend(text.trim())
                    text = ""
                }
                .padding(horizontal = 14.dp, vertical = 8.dp),
        ) {
            ChatText("发送", color = Color.White, fontSize = 14.sp)
        }
    }
}

File: sdk/ios/Sources/ServifyKit/UI/ChatPanel.swift
#if canImport(UIKit)
import SwiftUI
import Combine

/// 主题常量（Kotlin ChatThemeDefaults 镜像，色值逐项一致；D9 零三方依赖——
/// SwiftUI 全系统库，白名单约束由 Package.swift 无 dependencies 结构性保证）。
enum ChatThemeDefaults {
    /// ARGB UInt32 → SwiftUI Color（取低 24 位 RGB，alpha 恒 1——V1 无透明主色场景）。
    static func argb(_ value: UInt32) -> Color {
        Color(
            red: Double((value >> 16) & 0xFF) / 255.0,
            green: Double((value >> 8) & 0xFF) / 255.0,
            blue: Double(value & 0xFF) / 255.0
        )
    }

    /// 未配置 branding.primaryColor 时的默认主色（品牌蓝，0xFF2563EB）。
    static let defaultPrimary: UInt32 = 0xFF2563EB
    static let pageBackground = argb(0xFFF8FAFC)
    static let surfaceWhite = Color.white
    static let textPrimary = argb(0xFF0F172A)
    static let textSecondary = argb(0xFF64748B)
    static let bubbleCustomer = argb(0xFF2563EB)
    static let bubbleOther = Color.white
    static let divider = argb(0xFFE2E8F0)
    static let handoffBadge = argb(0xFFFEF3C7)

    static func resolvePrimary(_ configured: UInt32?) -> Color {
        argb(configured ?? defaultPrimary)
    }
}

/**
 * 面板状态接线（Kotlin 镜像：ChatPanel.kt 的 LaunchedEffect/collectAsState 面）：
 * 门面事件流 → @Published。合并规则不在本层——由 ChatPanelState（Linux 单测）
 * 与门面 historySnapshot（同 id 覆盖累积）承担，这里只做线程搬运与驱动。
 */
@MainActor
final class ChatPanelModel: ObservableObject {
    @Published private(set) var items: [ChatListItem] = []
    @Published private(set) var statusLine: String?
    @Published private(set) var connectionState: ConnectionState = .idle

    private let chat: ServifyChat
    private let panelState: ChatPanelState

    init(chat: ServifyChat, welcomeText: String?) {
        self.chat = chat
        self.panelState = ChatPanelState(welcomeText: welcomeText)
    }

    /// 面板挂载即启动（View 的 .task 调用一次）：
    /// 会话页可见清未读（§4.3）+ 回放门面累积（hide 期间消息不丢，D7 内存级连续性）
    /// + 订阅四条事件流。
    func start() {
        chat.onSessionVisible()
        chat.historySnapshot().forEach { panelState.onMessage($0) }
        items = panelState.items()

        Task {
            for await message in chat.events.messages.makeStream() {
                panelState.onMessage(message)
                items = panelState.items()
            }
        }
        Task {
            for await state in chat.events.connectionState.makeStream() { connectionState = state }
        }
        Task {
            for await message in chat.events.waitingInQueue.makeStream() { statusLine = message }
        }
        Task {
            for await assignment in chat.events.agentAssigned.makeStream() { statusLine = assignment.message }
        }
    }

    /// 发送客户消息；失败（echo 超时）经 error 流暴露给宿主，面板内静默可重发。
    func send(_ text: String) {
        Task { try? await chat.sendMessage(text) }
    }

    /// V1：协议无客户端转人工帧——点击即发送固定文本消息（服务端兜底链路可达）。
    func requestHandoff() {
        send("请转人工客服")
    }
}

/**
 * 会话面板（D8 三形态的抽屉/全屏共体：形态差异只在宿主容器高度，主体组件同一棵）。
 */
struct ChatPanel: View {
    let chat: ServifyChat
    let title: String
    let welcomeText: String?
    var onDismiss: () -> Void

    @StateObject private var model: ChatPanelModel
    @State private var inputText = ""

    init(chat: ServifyChat, title: String, welcomeText: String?, onDismiss: @escaping () -> Void) {
        self.chat = chat
        self.title = title
        self.welcomeText = welcomeText
        self.onDismiss = onDismiss
        _model = StateObject(wrappedValue: ChatPanelModel(chat: chat, welcomeText: welcomeText))
    }

    var body: some View {
        VStack(spacing: 0) {
            titleBar
            if let line = model.statusLine { statusLineView(line) }
            messageList
            inputBar(enabled: model.connectionState == .connected)
        }
        .background(ChatThemeDefaults.pageBackground)
        .task { model.start() }
    }

    private var primary: Color {
        ChatThemeDefaults.resolvePrimary(chat.config.branding.primaryColor)
    }

    // MARK: - 标题栏（主色底 + 标题 + 连接状态 + 关闭）

    private var titleBar: some View {
        ZStack {
            HStack {
                Text(title)
                    .font(.system(size: 17, weight: .semibold))
                    .foregroundColor(.white)
                    .lineLimit(1)
                Spacer()
                Button(action: onDismiss) {
                    Text("✕")
                        .font(.system(size: 18))
                        .foregroundColor(.white)
                        .padding(6)
                }
            }
            if let label = connectionLabel(model.connectionState) {
                Text(label)
                    .font(.system(size: 12))
                    .foregroundColor(.white.opacity(0.85))
                    .lineLimit(1)
            }
        }
        .padding(.horizontal, 16)
        .frame(height: 52)
        .background(primary)
    }

    private func connectionLabel(_ state: ConnectionState) -> String? {
        switch state {
        case .idle: return nil
        case .connecting: return "连接中…"
        case let .reconnecting(attempt): return "重新连接中…（第 \(attempt) 次）"
        case .connected: return nil
        case .disconnected: return "连接已断开"
        }
    }

    // MARK: - 状态条（转人工状态机渲染：排队提示 / 坐席接入提示，显示最近一条）

    private func statusLineView(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 13))
            .foregroundColor(ChatThemeDefaults.textPrimary)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
            .background(ChatThemeDefaults.handoffBadge)
    }

    // MARK: - 消息列表（新消息或流式增量到达即滚到底）

    private var messageList: some View {
        ScrollViewReader { proxy in
            ScrollView {
                LazyVStack(spacing: 0) {
                    ForEach(model.items.indices, id: \.self) { index in
                        itemView(model.items[index])
                            .id(listItemKey(model.items[index]))
                    }
                }
                .padding(.vertical, 8)
            }
            .onChange(of: model.items) { _ in
                if let last = model.items.last {
                    withAnimation {
                        proxy.scrollTo(listItemKey(last), anchor: .bottom)
                    }
                }
            }
        }
    }

    private func listItemKey(_ item: ChatListItem) -> String {
        switch item {
        case .welcome: return "__welcome"
        case let .bubble(bubble): return bubble.id
        }
    }

    @ViewBuilder
    private func itemView(_ item: ChatListItem) -> some View {
        switch item {
        case let .welcome(text): welcomeItem(text)
        case let .bubble(bubble): messageBubble(bubble)
        }
    }

    private func welcomeItem(_ text: String) -> some View {
        Text(text)
            .font(.system(size: 13))
            .foregroundColor(ChatThemeDefaults.textSecondary)
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(ChatThemeDefaults.surfaceWhite)
            .cornerRadius(12)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(.horizontal, 16)
            .padding(.vertical, 8)
    }

    private func messageBubble(_ item: ChatBubble) -> some View {
        let right = item.sender.isRightAligned
        return HStack {
            if right { Spacer(minLength: 60) }
            VStack(alignment: right ? .trailing : .leading, spacing: 0) {
                if item.isAiResponse && !right {
                    Text("AI 助手")
                        .font(.system(size: 11))
                        .foregroundColor(ChatThemeDefaults.textSecondary)
                        .padding(.bottom, 2)
                        .padding(.leading, 4)
                }
                Text(item.isStreaming ? item.content + "…" : item.content)
                    .font(.system(size: 15))
                    .foregroundColor(right ? .white : ChatThemeDefaults.textPrimary)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 8)
                    .background(right ? ChatThemeDefaults.bubbleCustomer : ChatThemeDefaults.bubbleOther)
                    .cornerRadius(12)
                    .frame(maxWidth: 280, alignment: right ? .trailing : .leading)
                if item.suggestsHandoff {
                    handoffButton
                }
                if !item.sources.isEmpty {
                    sourcesList(item.sources)
                }
            }
            if !right { Spacer(minLength: 60) }
        }
        .padding(.horizontal, 16)
        .padding(.vertical, 4)
    }

    /// 置信门"转人工"强调按钮（D8：next_action=handoff 驱动）。
    private var handoffButton: some View {
        Button {
            model.requestHandoff()
        } label: {
            Text("转人工")
                .font(.system(size: 13, weight: .medium))
                .foregroundColor(.white)
                .padding(.horizontal, 12)
                .padding(.vertical, 6)
                .background(primary)
                .cornerRadius(8)
        }
        .padding(.top, 4)
    }

    /// 引用来源列表（D8：按得分降序的展开列表，score 已在映射层排好、null 归底）。
    private func sourcesList(_ sources: [SourceItem]) -> some View {
        VStack(alignment: .leading, spacing: 2) {
            Text("参考来源")
                .font(.system(size: 11))
                .foregroundColor(ChatThemeDefaults.textSecondary)
            ForEach(sources.indices, id: \.self) { index in
                Text(sourceLine(sources[index]))
                    .font(.system(size: 12))
                    .foregroundColor(ChatThemeDefaults.textSecondary)
                    .lineLimit(1)
            }
        }
        .padding(.top, 4)
    }

    private func sourceLine(_ source: SourceItem) -> String {
        var line = "· \(source.title)（\(source.documentId)"
        if let score = source.score {
            line += String(format: "%.2f", score)
        }
        return line + "）"
    }

    // MARK: - 输入栏

    private func inputBar(enabled: Bool) -> some View {
        HStack(spacing: 8) {
            TextField("输入消息…", text: $inputText)
                .font(.system(size: 15))
                .foregroundColor(ChatThemeDefaults.textPrimary)
                .padding(.horizontal, 12)
                .padding(.vertical, 8)
                .background(ChatThemeDefaults.pageBackground)
                .cornerRadius(20)
            Button {
                let text = inputText.trimmingCharacters(in: .whitespacesAndNewlines)
                guard !text.isEmpty else { return }
                inputText = ""
                model.send(text)
            } label: {
                Text("发送")
                    .font(.system(size: 14))
                    .foregroundColor(.white)
                    .padding(.horizontal, 14)
                    .padding(.vertical, 8)
                    .background(canSend(enabled: enabled) ? primary : ChatThemeDefaults.divider)
                    .cornerRadius(20)
            }
            .disabled(!canSend(enabled: enabled))
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
        .background(ChatThemeDefaults.surfaceWhite)
    }

    private func canSend(enabled: Bool) -> Bool {
        enabled && !inputText.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }
}
#endif
